### DISCUSSION TITLE
Relax 'this' keyword restriction in field initializers

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#6128] [Language Ideas]
Is it possible to allow accessing other fields within field initializers, using `this`?
```cs
class Foo
{
  int _x = 1;
  int _y = this._x + 1;
}
```
Calling methods (and properties) would still be prohibited, only field access would be allowed. I think it should work because similar access is allowed in static context:
```cs
class Bar
{
  static int _x = 1;
  static int _y = _x + 1;
}
```

**Comment by HaloFour** [comment]
IIRC it's the runtime that forbids accessing `this` before the base constructor is called (which is when initializers are executed).

**Reply by svick** [reply]
[The CLI spec §III.1.8.1.4 *Class and object initialization rules* says:](https://www.ecma-international.org/wp-content/uploads/ECMA-335_6th_edition_june_2012.pdf#%5B%7B%22num%22%3A3210%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C87%2C498%2C0%5D)

> An object constructor shall not return unless a constructor for the base class or a different constructor for the object’s class has been called on the newly constructed object. The verification algorithm shall treat the `this` pointer as uninitialized unless the base class constructor has been called. No operations can be performed on an uninitialized `this` except for storing into and loading from the object’s fields.

Since accessing fields (but only accessing fields) is allowed before the base constructor is called, C# could theoretically allow this in instance field initializers.

**Reply by HaloFour** [reply]
Yes, if you're permitted to load from the object's fields that would definitely make this possible.

I would have concerns about the ordering of assignments, perhaps the compiler could enforce assignments in lexical order, and not permit referencing fields in other `partial` declarations?

**Reply by TahirAhmadov** [reply]
> ordering of assignments

That's why I mentioned static fields: the same logic can be applied to instance fields.

**Reply by HaloFour** [reply]
> That's why I mentioned static fields: the same logic can be applied to instance fields.

Given the compiler doesn't enforce anything here it's very easy to reference a field that has not been initialized.  I've been bitten by this, and I don't think that the compiler should enable new scenarios where it is possible.

**Reply by TahirAhmadov** [reply]
> compiler doesn't enforce anything here

Oops, I was under the impression that it does. This is unfortunate as it creates a dilemma: either replicate the same behavior of not enforcing the ordering, which would be consistent, or come up with a topological sort mechanism, which would be more reliable, but inconsistent with static fields.
I'm wondering if this question was discussed in detail back when this came up for static fields. I have a feeling it was decided not to enforce ordering for a reason back then, and there is high chance the same reason would apply to instance fields.

**Reply by CyrusNajmabadi** [reply]
Do it in lexical order in the file. Disallow across files.

**Reply by TahirAhmadov** [reply]
Does lexical order mean "sorted by name"? Or "sorted by placement"?
If it's by name, this will be problematic - you have to name your fields in a way that reflects their dependencies.

**Reply by CyrusNajmabadi** [reply]
Sorted by position in the file

**Reply by TahirAhmadov** [reply]
Is accessing across files allowed in static fields?

**Reply by CyrusNajmabadi** [reply]
yes.  i would not want to make that mistake again.

**Reply by TahirAhmadov** [reply]
What do you think about topological sort by dependency? That way placement doesn't matter and across files is allowed.

**Reply by CyrusNajmabadi** [reply]
i will never ever support this.  There's nothing in the language that makes me think data dependency should have any influence on this sort of thing.

**Reply by TahirAhmadov** [reply]
I think about it in a similar way to C# vs C++ WRT "includes". Whereas C++ needs the definition to be "included" to be available, in C# the compiler is able to link all the identifiers without it; basically all identifiers are available everywhere (with the right access). Similarly, instead of relying on lexical order, it would be cool if C# could basically put all the fields into a bucket and make them available to within the class.

But for me personally, it's not a deal breaker.

**Comment by CyrusNajmabadi** [comment]
Definitely an interesting possibility.

**Comment by TahirAhmadov** [comment]
Some other discussions I found which are related:
https://github.com/dotnet/csharplang/discussions/5813
https://github.com/dotnet/csharplang/discussions/5845

Along with allowing to access other fields in field initializers, I think it should also be allowed to access other fields when calling other constructors, like this: `: this(this._field)` or `: base(this._field)`.

**Comment by bbartels** [comment]
I come across this issue relatively often with primary constructors, as there is no way to store a temporary value as part of a computation from a constructor parameter. Wanting to set two fields to the same value as a result of an operation in the constructor. For instance:
```csharp
public class SomeService(TimeProvider timeProvider)
{
    private readonly DateTimeOffset StartTime = timeProvider.GetUtcNow();
    private DateTimeOffset TimeEventLastSeen = StartTime;
}
```

Lexical order, not across files sounds like a very intuitive way of relaxing this restriction.

**Reply by AndersKehlet** [reply]
This is how F# works. With primary constructors we effectively treat the body of the type as the scope of the constructor. It makes a ton of sense then to also allow simple assignments in lexical order.

**Reply by RikkiGibson** [reply]
#6591 suggests allowing a "final initializer" that runs after the constructor returns, but before the instance is used at the creation site. That might address this scenario.

**Reply by AndersKehlet** [reply]
In C# we'd probably need something like init for more complex scenarios with temp vars, but I'd still want this feature. Roslyn emits a constructor for these assignments anyway.
